tests/integration/v3_watch_test.go

```diff
@@ -1397,3 +1397,65 @@ func TestV3WatchCloseCancelRace(t *testing.T) {
 		t.Fatalf("expected %s watch, got %s", expected, minWatches)
 	}
 }
+
+// TestV3WatchProgressWaitsForSync checks that progress notifications
+// don't get sent until the watcher is synchronised
+func TestV3WatchProgressWaitsForSync(t *testing.T) {
+	integration.BeforeTest(t)
+
+	clus := integration.NewCluster(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	client := clus.RandClient()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	// Write a couple values into key to make sure there's a
+	// non-trivial amount of history.
+	count := 100
+	t.Log(fmt.Sprintf("Writing key 'foo' %d times", count))
+	for i := 0; i < count; i++ {
+		_, err := client.Put(ctx, "foo", fmt.Sprintf("bar%d", i))
+		require.NoError(t, err)
+	}
+
+	// Create watch channel starting at revision 1 (i.e. it starts
+	// unsynced because of the update above)
+	wch := client.Watch(ctx, "foo", clientv3.WithRev(1))
+
+	// Immediately request a progress notification. As the client
+	// is unsynchronised, the server will have to defer the
+	// notification internally.
+	err := client.RequestProgress(ctx)
+	require.NoError(t, err)
+
+	// Verify that we get the watch response first
+	wr := <-wch
+	if wr.Err() != nil {
+		t.Fatal(fmt.Errorf("watch error: %w", wr.Err()))
+		return
+	}
+	if wr.IsProgressNotify() {
+		t.Fatal("Progress notification from unsynced client!")
+		return
+	}
+	if len(wr.Events) != count || wr.Header.Revision != int64(count+1) {
+		t.Fatal("Incomplete watch response!")
+		return
+	}
+
+	// ... followed by the requested progress notification
+	wr2 := <-wch
+	if wr2.Err() != nil {
+		t.Fatal(fmt.Errorf("watch error: %w", wr.Err()))
+		return
+	}
+	if !wr2.IsProgressNotify() {
+		t.Fatal("Did not receive progress notification!")
+		return
+	}
+	if wr2.Header.Revision != int64(count+1) {
+		t.Fatal("Wrong revision in progress notification!")
+		return
+	}
+}
```

Comment:
The test case isn't clear to me. I suggest to create two goroutines, one keeps calling RequestProgress and writing data, and the other keeps checking the revision. 

The first goroutine:
```
go func() {
    defer func() {
        time.Sleep(10*time.Millisecond)
        cancel()
    }()
    for i:=0; i<100; i++ {
        // RequestProgress before writing data
        err := client.RequestProgress(ctx)
        require.NoError(t, err)

        _, err = client.Put(ctx, "foo", fmt.Sprintf("bar%d", i))
        require.NoError(t, err)

       // RequestProgress after writing data
        err := client.RequestProgress(ctx)
        require.NoError(t, err)

        time.Sleep(10*time.Millisecond)
    }
}()
```

The second goroutine:
```
go func() {
    wch := client.Watch(ctx, "foo")
    for resp := range wch {
        // Follow https://github.com/etcd-io/etcd/blob/c24ba4d3e85ef1147d9245f544c125e251ff03e8/tests/linearizability/watch.go#L101-L139
    }
}()

```

Reply:
Please add this test case as the first commit, and confirm that it can easily reproduce the original issue. Then apply the fix as the second commit, and confirm that you can't reproduce the issue any more.

Reply:
Your proposal is actually a very different test case, which would especially have a much lower likelihood of reproducing the problem. The point here is that we
1. Create a database state where a key has 100+ revisions of history (no watcher active at this point!)
2. Create a new watcher that starts at revision 1 (and therefore is 100 revisions "behind" at creation time)
3. Immediately ask for a progress notification. This is highly likely to catch the watcher in a state where it has not "caught up" yet.

Before the patch, this typically gives you:
1. Confirmation that worker was created
2. Progress notification (revision 101 in header)
3. Watcher notification for values 1-100

Because it's basically a race between the watch and the control stream, and given the complexity in generating the response for all 100 revisions, the control stream basically always wins. From the client's perspective, this leaves a very obvious window where a connection drop would cause 100 revisions of history to be lost.

The fix basically switches the order between answers 2 and 3 (by deferring the progress notification). I hope that makes sense? Any way I could document it better?

Reply:
Well, it's a race, so there's never 100%. According to a series of runs, with 100 revisions it reproduces roughly 99.5% for me. With 1000 revisions it's solidly > 99.99% now according to my test, so let's stay with that.